### DISCUSSION TITLE
Ceiling units changed in grid_stat?

### DISCUSSION THREAD

**Discussion by johnlwagner** [dtcenter/METplus#1744] [Statistical Computation] [answered]
Greetings
When I run grid_stat for ceiling verification, the units output by MET seems to have switched from meters to kilometers.  This seems to have begun since we transitioned from WCOSS1 to WCOSS2.  On WCOSS1, I believe we were using v10.0.0.  On WCOSS2, we are using v10.0.1.  I don't see anything in the v10.0.1 release notes that would cause this change.
I have looked at the metadata on all of our grib2 files.  They all have a 13 in octet 11 in section 4, which defines the data in meters.
This only seems to be an issue with grid_stat.  We process the same grib2 files with point_stat.  All of the output from point_stat continues to be in meters.
Is there a setting that I am missing that would change the units?  Is there a system setting being used that may have changed between WCOSS1 and WCOSS2?
I am still investigating this issue.  If I learn anything new, I will update this issue.
Thanks
John

**Comment by JohnHalleyGotway** [comment]
@johnlwagner I saw this new issue in the MET repo. Since it is really more of a question, I converted it into a METplus Discussion topic.

There certainly was no intentional change to the MET software that would cause the change in units you describe. In fact, MET, in general, DOES NOT convert units behind the scenes in any standard way. The only exception I can think of is that we do convert GRIB pressure levels from hectopascals to millibars to be consistent with outputs from wgrib and wgrib2. But there's no embedded unit conversion of the raw data itself.

Instead, the user can configure calls to the tools using the "convert(x)" command. That is usually specified to handle unit conversions. I'd start debugging this by checking 3 things:

1. Has the data steam actually changed? Pull two sample files from WCOSS1 (if it's still available?) and WCOSS2 that should be the same. Run wgrib2 with -V option to print the min/max data values. Do the range of values match? You could also run them through plot_data_plane to see if the plotting range matches.

2. Perhaps you were already using the "convert(x)" function in your runtime scripts, but that logic has broken for some reason on WCOSS2? You could just grep for "convert" in your shell scripts.

3. Perhaps your configuration has remained the same, but MET is matching a different GRIB record than it did previously? This would require some more investigation and might be a bug on the MET side.

Hopefully that helps.

**Comment by johnlwagner** [comment]
It appears that my issue may be with regrid_data_plane, not grid_stat.  The grib2 input file is in meters.  The nc output file is in km.
On cactus, my test case is in /lfs/h1/mdl/ptmp/john.l.wagner/verif/test1/matching_urma_afvt_2022061500_co_cg_grid.248846/NDFDproj/006.  My regrid_data_plane command is 
/apps/ops/para/libs/intel/19.1.3.304/met/10.0.1/bin/regrid_data_plane blendcogb2p522061416cgf008.grb2 /lfs/h1/mdl/verif/noscrub/usr/john.l.wagner/mdl-verification/verif_v2.1.0/fix/urma_mask/blend_co blendconc2p522061416cgf008 -method NEAREST -width 1 -v 0 -name "cg" -field "name=\"CEIL\"; level=\"L0\";"
I ran this same command with v9.1.3 and got the same result.  I'm not sure if this is a MET issue or a system issue.  Its certainly not a MET version issue.

I take back what I said about this not being an issue with point_stat as well.  I have found instances where this is happening with that output as well.  This makes sense, as we use regrid_data_plane before calling both grid_stat and point_stat.

1.  I did compare data from WCOSS1 (which was rsynced from WCOSS1 to WCOSS2) and data that was archived on WCOSS2.  They have the same min value (30), similar max values (67590.9 vs 64972.6), and are both listed in meters.
2. I am not aware of us using convert(x) anywhere in our scripts.
3. I don't think the record being used has changed.

**Reply by JohnHalleyGotway** [reply]
@johnlwagner my WCOSS2 access isn't working right now, but I do see that you're using URMA data. So I pulled this sample URMA file from NOMADS:
```
wget https://nomads.ncep.noaa.gov/pub/data/nccf/com/urma/prod/urma2p5.20220807/urma2p5.t12z.2dvaranl_ndfd.grb2_wexp
```
I see that it does indeed contain CEIL:
```
wgrib2 urma2p5.t12z.2dvaranl_ndfd.grb2_wexp | grep CEIL | wgrib2 -i -V urma2p5.t12z.2dvaranl_ndfd.grb2_wexp
```
Here's the wgrib2 output:
```
12:67879563:vt=2022080712:cloud ceiling:anl:CEIL Ceiling [m]:
    ndata=3744965:undef=0:mean=11652.6:min=1:max=16000
```
To test, I ran `regrid_data_plane` to basically extract CEIL and regrid it to the grid it's already on. So the input and output should be identical:
```
regrid_data_plane urma2p5.t12z.2dvaranl_ndfd.grb2_wexp urma2p5.t12z.2dvaranl_ndfd.grb2_wexp CEIL.nc -field 'name="CEIL"; level="L0";'
```
And the log messages report the same range of input/output values:
```
DEBUG 2: Range of input data (name="CEIL"; level="L0";) is 1 to 16000.
DEBUG 2: Range of regridded data (name="CEIL"; level="L0";) is 1 to 16000.
```
So you can see that it's NOT actually changing units of the data. And checking the header of the NetCDF output file (`ncdump -h CEIL.nc`) I see the units reported as meters:
```
	float CEIL_L0(lat, lon) ;
		CEIL_L0:name = "CEIL_L0" ;
		CEIL_L0:long_name = "Ceiling" ;
		CEIL_L0:level = "L0" ;
		CEIL_L0:units = "m" ;
```
Please note though that all MET is doing here is a GRIB2 table lookup. Whatever units string it finds as a match in that GRIB2 table are just copied into the units attribute in the output.

So perhaps this is more of an issue in the handling of the metadata instead of the actual data values themselves? I don't see any issue with this sample data file I'm using... wgrib2 and MET both agree that the units are meters. So the default GRIB2 tables are working fine for MET.

Do you see different behavior with the file on cactus? If so, it could be related to your use of the $MET_GRIB_TABLES environment variable. Users can use that to override the default GRIB tables that are used. If your scripts are setting that environment variable, that'd be a potential source of the issue to check.

If you'd like me to look more closely at your specific GRIB2 file, please post it to our anonymous FTP site following the instructions in dtcenter/METplus#954.

**Reply by johnlwagner** [reply]
I repeated your steps.  I see the same results as you except for when I ncdump -h CEIL.nc, the units are listed as km.  When I look at the actual nc data, the maxval I see is 16000, which implies that the data is indeed in meters and that the label in the ncdump header is wrong.  Perhaps there is a change in the nc library between WCOSS1 and WCOSS2?

Here are some sample grid_stat output from WCOSS1 (bottom; m) and WCOSS2 (top; km):
cg_afvt/grid_stat_080000L_20220615_000000V.stat:V10.0.1 BLEND NA   080000    20220615_000000 20220615_000000 000000   20220615_000000 20220615_000000 cg       km         L0       cg      km        L0      URMA   FULL         NEAREST     1           <=60.96,<=152.4,<=304.8,<=914.4  <=60.96,<=152.4,<=304.8,<=914.4  NA         NA    MCTC      2953665      5             450               160                 231             593         603           9642          18173          25471 5030         22201        1716        7273          56312 12507 13851         3145        6201       37602         42617          95242          14344        11518         28864       76306       2463613
cg_afvt.meters/grid_stat_080000L_20220615_000000V.stat:V10.0.0 BLEND NA   080000    20220615_000000 20220615_000000 000000   20220615_000000 20220615_000000 cg       m          L0       cg      m         L0      URMA   FULL         NEAREST     1           <=60.96,<=152.4,<=304.8,<=914.4  <=60.96,<=152.4,<=304.8,<=914.4  NA         NA    MCTC      2953665      5             424               145                 223             581         593           8798          14314          23032 5068          23453        2076        8757          60684 15432 20254         2408        4120       26455          34471          83146          15591        15989         38086       81501       2468064

The counts are certainly different, but they are on the same order of magnitude.  If the units were changed from m to km, I would not expect that to be the case.  To test this, I changed the units in my grid_stat config file from m to km.  Here are the new results:
cg_afvt/grid_stat_080000L_20220615_000000V.stat:V10.0.1 BLEND NA   080000    20220615_000000 20220615_000000 000000   20220615_000000 20220615_000000 cg       km         L0       cg      km        L0      URMA   FULL         NEAREST     1           <=.06096,<=.1524,<=.3048,<=.9144    <=.06096,<=.1524,<=.3048,<=.9144    NA         NA    MCTC      2953665     5           0                 0                   0               0           0              0              0           0  0     0           0           0        0  0    0          0           0        0  0  0           0         0          0  0       2953665
cg_afvt.meters/grid_stat_080000L_20220615_000000V.stat:V10.0.0 BLEND NA   080000    20220615_000000 20220615_000000 000000   20220615_000000 20220615_000000 cg       m          L0       cg      m         L0      URMA   FULL         NEAREST     1           <=60.96,<=152.4,<=304.8,<=914.4  <=60.96,<=152.4,<=304.8,<=914.4  NA         NA    MCTC      2953665      5             424               145                 223             581         593           8798          14314          23032 5068          23453        2076        8757          60684 15432 20254         2408        4120       26455          34471          83146          15591        15989         38086       81501       2468064

I guess the real issue is that the units reported by grid_stat are km and not m.  This is causing an issue when we run stat_analysis and it combines stat files from WCOSS1 and WCOSS2.  I'm not sure I can trust the results based on the results that I'm seeing.
It seems that the best course of action on my part is to rerun grid_stat and point_stat on WCOSS2 for any cases that were previously run on WCOSS1. 
I'm going to open a ticket with the WCOSS2 help desk to see if they can explain the unit change in the ncdump header.

**Reply by johnlwagner** [reply]
FYI I found the same issue with our wind direction verification.  The unit "degtrue" from WCOSS1 is now "degrees" on WCOSS2.

**Reply by JohnHalleyGotway** [reply]
@johnlwagner, thanks for running through those same steps.

This really has nothing to do with the version of the NetCDF library. The units reported by MET (both in the NetCDF headers for gridded data and the units column of the .stat output files) are simply pulled from the GRIB2 table files which are defined in the MET code base.

I strongly suspect that we're getting different GRIB tables on WCOSS1 vs WCOSS2. Please check to see if your scripts set a MET_GRIB_TABLES environment variable. If so, that'd imply that it's using a custom GRIB table that'd override those defaults. My guess is that you were using MET_GRIB_TABLES on WCOSS1 but did not copy over the custom table when you moved to GRIB2. At least, that'd be consistent with the behavior you describe.

Please LMK what you find.

**Comment by johnlwagner** [comment]
I confirmed that we are not using MET_GRIB_TABLES.  I don't believe we have
ever set that variable.
I'll continue to look into other wgrib settings on WCOSS2 that may be
different from WCOSS1.

On Fri, Aug 12, 2022 at 5:31 PM johnhg ***@***.***> wrote:

> @johnlwagner <https://github.com/johnlwagner>, thanks for running through
> those same steps.
>
> This really has nothing to do with the version of the NetCDF library. The
> units reported by MET (both in the NetCDF headers for gridded data and the
> units column of the .stat output files) are simply pulled from the GRIB2
> table files which are defined in the MET code base.
>
> I strongly suspect that we're getting different GRIB tables on WCOSS1 vs
> WCOSS2. Please check to see if your scripts set a MET_GRIB_TABLES
> environment variable. If so, that'd imply that it's using a custom GRIB
> table that'd override those defaults. My guess is that you were using
> MET_GRIB_TABLES on WCOSS1 but did not copy over the custom table when you
> moved to GRIB2. At least, that'd be consistent with the behavior you
> describe.
>
> Please LMK what you find.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dtcenter/METplus/discussions/1744#discussioncomment-3387416>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGZ7TYOQUZU3KGIRO53B4ZDVY27BVANCNFSM56MDMLUA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
John Wagner
Verification Task Lead
NOAA/National Weather Service
Meteorological Development Laboratory
Digital Forecast Services Division
SSMC2 Room 10106
Silver Spring, MD 20910
(240) 622-9138 <https://mail.google.com/mail/u/0/#m_226873644979562856_>
(Google
Voice)
(301) 427-9471 (office)
(908) 902-4155 (cell/text)

**Reply by JohnHalleyGotway** [reply]
@johnlwagner I'm just following up on GitHub discussions that are a bit old to see if we can mark them as being answered. I see that you are NOT using custom GRIB tables by setting the $MET_GRIB_TABLES environment variable. Have you found anything else that we explain the discrepancy in units reported on WCOSS vs WCOSS2?

**Comment by johnlwagner** [comment]
I have not found a reason yet as to why the units are changing.  I opened a
ticket with the WCOSS2 help desk.  They are not aware of anything that
would be causing this on the system.  They did recommend a cdo command to
change the units after regridding.

On Tue, Aug 23, 2022 at 12:59 PM johnhg ***@***.***> wrote:

> @johnlwagner <https://github.com/johnlwagner> I'm just following up on
> GitHub discussions that are a bit old to see if we can mark them as being
> answered. I see that you are NOT using custom GRIB tables by setting the
> $MET_GRIB_TABLES environment variable. Have you found anything else that we
> explain the discrepancy in units reported on WCOSS vs WCOSS2?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dtcenter/METplus/discussions/1744#discussioncomment-3458662>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGZ7TYM3AWBS2VXEEU4SED3V2T7PTANCNFSM56MDMLUA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
John Wagner
Verification Task Lead
NOAA/National Weather Service
Meteorological Development Laboratory
Digital Forecast Services Division
SSMC2 Room 10106
Silver Spring, MD 20910
(240) 622-9138 <https://mail.google.com/mail/u/0/#m_226873644979562856_>
(Google
Voice)
(301) 427-9471 (office)
(908) 902-4155 (cell/text)

**Reply by JohnHalleyGotway** [reply]
@jprestop since my WCOSS2 access is currently not working, I'm wondering if you'd be willing to do a screen sharing session with me to log on, replicate the behavior @johnlwagner reports, and investigate further?

```
# Load the METplus modules
wget https://nomads.ncep.noaa.gov/pub/data/nccf/com/urma/prod/urma2p5.20220807/urma2p5.t12z.2dvaranl_ndfd.grb2_wexp
regrid_data_plane urma2p5.t12z.2dvaranl_ndfd.grb2_wexp urma2p5.t12z.2dvaranl_ndfd.grb2_wexp CEIL.nc -field 'name="CEIL"; level="L0";'
ncdump -h CEIL.nc | grep units
```
Running on my machine, the units are reported as `m` for meters while @johnlwagner is seems `km` for kilometers. If we can replicate the behavior, we should be able to figure out why it is happening.

**Reply by jprestop** [reply]
@JohnHalleyGotway I'd be happy to do a screen sharing session.  Please feel free to schedule something on my Google calendar at a time that works well for you.

**Comment by JohnHalleyGotway** [comment, marked as answer]
@johnlwagner good news. @jprestop and I met today to debug behavior on WCOSS2 and now understand what is happening. We have short-term bandaid to recommend and a longer term solution that I need to write up as a new GitHub issue in MET.

Short-Term Bandaid
--------------------
You actually have 2 choices for this. Either add `GRIB2_cntr = 0;` or add `set_attr_units = "m";` to your configuration strings for ceiling.

The first option gives the MET library code more information to select the correct GRIB2 table entry for your data:
```
name = "CEIL"; level = "L0"; GRIB2_cntr = 0;
```
The second option lets the MET library code read the wrong GRIB2 table entry for your data, but then manually overrides it with the string you tell it to use:
```
name = "CEIL"; level = "L0"; set_attr_units = "m";
```
Both approaches achieve the desired result, although @jprestop and I both prefer using `GRIB2_cntr = 0;`.

Underlying Problem
-------------------
The MET GRIB2 tables contain 2 entries for the "CEIL" variable. One where the USAF (center 57) uses "km" and once where NOAA (center 0?) uses "m":
```
grib2_all.txt:0 1 0 255 0 0 6 13  "CEIL"  "Ceiling"  "m"
grib2_gpp_af.txt:0 14 0 25 57 0 6 13 "CEIL" "Ceiling" "km"
```
On WCOSS1 and when using MET version 10.1.0 on WCOSS2, @jprestop and I found that first entry listed is used. And that results in the desired result of units = "m". However when running with version 10.0.1 on WCOSS2, the second entry is used, which results in units = "km".
But why? The MET library code reads GRIB2 table information from files matching "data/table_files/grib2*". But there is no guarantee on the order in which those files are served up to be read. We could FORCE an order by alphabetizing them, but that'd just be an insufficient patch as well. So for whatever reason, they're read in one order for v10.0.1 of MET and a different order for v10.1.0! And that explains the odd behavior you're seeing. In fact, this info is printed to the screen at DEBUG level 3 or higher (e.g. `-v 3`):
```
DEBUG 3: Multiple GRIB2 table entries match lookup criteria (parm_name = CEIL):
DEBUG 3:   parm_name: CEIL, index_a = 0, grib2_mtab = 1, grib2_cntr = 0, grib2_ltab = 0, index_b = 6, index_c = 13
DEBUG 3:   parm_name: CEIL, index_a = 0, grib2_mtab = 14, grib2_cntr = 57, grib2_ltab = 0, index_b = 6, index_c = 13
```
Those log messages will appear in one order for v10.0.1 and reverse order for v10.1.0.

What's the Real Fix?
------------------
Right now, the MET library code reads `name="CEIL";` and searches for a matching GRIB2 table entry. If it finds multiples, it just uses the first match found. And it stores the "units" and "long name" from the GRIB2 table entry.

But that logic isn't insufficient! We can't really know which table entry to use until we actually read the GRIB2 records themselves. So we need to modify the logic so that we don't store the "units" and "long name" until we actually read the data.

This issue has actually existed for a very long time. It just hasn't bubbled up yet since *most* of the time, the same units are used for the same variable names across centers. In this case, the USAF apparently stores ceiling in kilometers while NOAA uses meters.

And here's the corresponding MET development issue: dtcenter/MET#2240
Thanks for your patience in unravelling this mystery.

**Comment by johnlwagner** [comment]
Thanks for the great detective work!  It is much appreciated.
We will go ahead and set the GRIB2_cntr as you suggested.  I like this
solution as well, as we won't have to set up anything to be
element-specific in our scripts.
Thanks again!
John

On Wed, Aug 24, 2022 at 5:24 PM johnhg ***@***.***> wrote:

> @johnlwagner <https://github.com/johnlwagner> good news. @jprestop
> <https://github.com/jprestop> and I met today to debug behavior on WCOSS2
> and now understand what is happening. We have short-term bandaid to
> recommend and a longer term solution that I need to write up as a new
> GitHub issue in MET.
> Short-Term Bandaid
>
> You actually have 2 choices for this. Either add GRIB2_cntr = 0; or add set_attr_units
> = "m"; to your configuration strings for ceiling.
>
> The first option gives the MET library code more information to select the
> correct GRIB2 table entry for your data:
>
> name = "CEIL"; level = "L0"; GRIB2_cntr = 0;
>
> The second option lets the MET library code read the wrong GRIB2 table
> entry for your data, but then manually overrides it with the string you
> tell it to use:
>
> name = "CEIL"; level = "L0"; set_attr_units = "m";
>
> Both approaches achieve the desired result, although @jprestop
> <https://github.com/jprestop> and I both prefer using GRIB2_cntr = 0;.
> Underlying Problem
>
> The MET GRIB2 tables contain 2 entries for the "CEIL" variable:
>
> grib2_all.txt:0 1 0 255 0 0 6 13  "CEIL"  "Ceiling"  "m"
> grib2_gpp_af.txt:0 14 0 25 57 0 6 13 "CEIL" "Ceiling" "km"
>
> On WCOSS1 and when using MET version 10.1.0 on WCOSS2, @jprestop
> <https://github.com/jprestop> and I found that first entry listed is
> used. And that results in the desired result of units = "m". However when
> running with version 10.0.1 on WCOSS2, the second entry is used, which
> results in units = "km".
> But why? The MET library code reads GRIB2 table information from files
> matching "data/table_files/grib2*". But there is no guarantee on the order
> in which those files are served up to be read. We could FORCE an order by
> alphabetizing them, but that'd just be an insufficient patch as well. So
> for whatever reason, they're read in one order for v10.0.1 of MET and a
> different order for v10.1.0! And that explains the odd behavior you're
> seeing. In fact, this info is printed to the screen at DEBUG level 3 or
> higher (e.g. -v 3):
>
> DEBUG 3: Multiple GRIB2 table entries match lookup criteria (parm_name = CEIL):
> DEBUG 3:   parm_name: CEIL, index_a = 0, grib2_mtab = 1, grib2_cntr = 0, grib2_ltab = 0, index_b = 6, index_c = 13
> DEBUG 3:   parm_name: CEIL, index_a = 0, grib2_mtab = 14, grib2_cntr = 57, grib2_ltab = 0, index_b = 6, index_c = 13
>
> Those log messages will appear in one order for v10.0.1 and reverse order
> for v10.1.0.
> What's the Real Fix?
>
> Right now, the MET library code reads name="CEIL"; and searches for a
> matching GRIB2 table entry. If it finds multiples, it just uses the first
> match found. And it stores the "units" and "long name" from the GRIB2 table
> entry.
>
> But that logic isn't insufficient! We can't really know which table entry
> to use until we actually read the GRIB2 records themselves. So we need to
> modify the logic so that we don't store the "units" and "long name" until
> we actually read the data.
>
> This issue has actually existed for a very long time. It just hasn't
> bubbled up yet since *most* of the time, the same units are used for the
> same variable names across centers. In this case, the USAF apparently
> stores ceiling in kilometers while NOAA uses meters.
>
> I'll write this up as an issue for MET.
>
> Thanks for you patience in unravelling this mystery.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dtcenter/METplus/discussions/1744#discussioncomment-3469365>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGZ7TYOWYUOBQJXA2VAIFXLV22HIXANCNFSM56MDMLUA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
John Wagner
Verification Task Lead
NOAA/National Weather Service
Meteorological Development Laboratory
Digital Forecast Services Division
SSMC2 Room 10106
Silver Spring, MD 20910
(240) 622-9138 <https://mail.google.com/mail/u/0/#m_226873644979562856_>
(Google
Voice)
(301) 427-9471 (office)
(908) 902-4155 (cell/text)
